src/connection.sml

```diff
@@ -1,68 +1,194 @@
-            val addr = INetSock.toAddr(NetHostDB.addr postgres, 5432)
-        in
-            Socket.connect (socket, addr);
-            socket
-        end
+structure PostgresClient =
+struct
 
-    fun convertString (elem: string) =
-        (Word8Vector.fromList (List.map Byte.charToByte (String.explode elem)))
+  datatype 't Operation = OK | ERROR of 't
 
-    fun intToBytes (n: int) =
-        Word8Vector.fromList
-            [ Word8.andb ((Word8.>> ((Word8.fromInt n), 0wx24)), 0wxFF),
-              Word8.andb ((Word8.>> ((Word8.fromInt n), 0wx16)), 0wxFF),
-              Word8.andb ((Word8.>> ((Word8.fromInt n), 0wx8)), 0wxFF),
-              Word8.andb (Word8.fromInt n, 0wxFF) ]
+  fun connect () =
+    let
+      val socket: Socket.active INetSock.stream_sock = INetSock.TCP.socket ()
+      val SOME postgres = NetHostDB.getByName "127.0.0.1"
+      val addr = INetSock.toAddr (NetHostDB.addr postgres, 5432)
+    in
+      Socket.connect (socket, addr);
+      socket
+    end
 
-    fun bytesToInt (bytes: Word8Vector.vector) =
-        let
-            open Word8
-            open Word8VectorSlice
-            val slices = full bytes
-            val first = sub(slices, 0)
-            val second = sub(slices, 1)
-            val third = sub(slices, 2)
-            val fourth = sub(slices, 3)
-        in
-            Word8.toInt (orb (fourth, (orb ((<< (third, 0wx8)), (orb ((<< (second, 0wx16)), (<< (first, 0wx24))))))))
-        end
+  fun convertString (elem: string) =
+    (Word8Vector.fromList (List.map Byte.charToByte (String.explode elem)))
 
-    fun startup (socket: Socket.active INetSock.stream_sock) =
-        let 
-            val message = convertString "user\000admin\000database\000socket\000\000"
-            val length = 
-                intToBytes 36
-                (* Word8Vector.fromList [ 0wx50, 0wx0, 0wx0, 0wx0 ] *)
-            val protocolVersion = 
-                Word8Vector.fromList [0w0, 0w3, 0w0, 0w0]
-            val buffer = Word8VectorSlice.full (Word8Vector.concat [ length, protocolVersion, message ])
-            val buf = PolyML.makestring buffer
-        in 
-            print buf;
-            print "\n\n";
+  fun intToBytes (n: int) =
+    Word8Vector.fromList
+      [ Word8.andb ((Word8.>> ((Word8.fromInt n), 0wx24)), 0wxFF)
+      , Word8.andb ((Word8.>> ((Word8.fromInt n), 0wx16)), 0wxFF)
+      , Word8.andb ((Word8.>> ((Word8.fromInt n), 0wx8)), 0wxFF)
+      , Word8.andb (Word8.fromInt n, 0wxFF)
+      ]
 
-            Socket.sendVec(socket, buffer)
-        end
+  fun bytesToInt (bytes: Word8Vector.vector) =
+    let
+      open Word8
+      open Word8VectorSlice
+      val slices = full bytes
+      val first = sub (slices, 0)
+      val second = sub (slices, 1)
+      val third = sub (slices, 2)
+      val fourth = sub (slices, 3)
+    in
+      Word8.toInt (orb (fourth, (orb ((<<(third, 0wx8)), (orb
+        ((<<(second, 0wx16)), (<<(first, 0wx24))))))))
```

Comment:
what's this << syntax?

Reply:
This is a bitwise shift: https://www.ibm.com/docs/en/i/7.2?topic=expressions-bitwise-left-right-shift-operators